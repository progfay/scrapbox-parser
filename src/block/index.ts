import { convertToCodeBlock } from "./CodeBlock";
import { convertToLine } from "./Line";
import { convertToTable } from "./Table";
import { convertToTitle } from "./Title";

import type { CodeBlock } from "./CodeBlock";
import type { Line } from "./Line";
import type { Pack } from "./Pack";
import type { Table } from "./Table";
import type { Title } from "./Title";

/**
 * Scrapbox block type
 */
export type Block = Title | CodeBlock | Table | Line;

export const convertToBlock = (pack: Pack): Block => {
	switch (pack.type) {
		case "title":
			return convertToTitle(pack);

		case "codeBlock":
			return convertToCodeBlock(pack);

		case "table":
			return convertToTable(pack);

		case "line":
			return convertToLine(pack);
	}
};
