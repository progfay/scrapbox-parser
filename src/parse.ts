import { BlockType, convertToBlock } from './block'
import { BlockComponentType, convertToBlockComponents } from './block/BlockComponent'
import { PackedBlockComponentType, packBlockComponents } from './block/PackedBlockComponent'

type PageType = {
  title: string
  blocks: BlockType[]
}

export const convertToBlocks = (blockComponents: BlockComponentType[]): BlockType[] => {
  const packedBlockComponents: PackedBlockComponentType[] = packBlockComponents(blockComponents)
  return packedBlockComponents.map(convertToBlock)
}

export const parse = (input: string): PageType => {
  const blockComponents: BlockComponentType[] = convertToBlockComponents(input.trim())

  const [firstBlock, ...body] = blockComponents
  const title = firstBlock && firstBlock.text ? firstBlock.text : 'Untitled'
  const blocks = convertToBlocks(body)

  return { title, blocks }
}
