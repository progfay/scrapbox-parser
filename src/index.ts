import { parse } from './parse'

export default parse
export { BlockType } from './block'
export { CodeBlockType } from './block/CodeBlock'
export { TableType } from './block/Table'
export { LineType } from './block/Line'
export { LineNodeType } from './block/node'
export { QuoteNodeType } from './block/node/QuoteNode'
export { HelpfeelNodeType } from './block/node/HelpfeelNode'
export { StrongImageNodeType } from './block/node/StrongImageNode'
export { StrongIconNodeType } from './block/node/StrongIconNode'
export { StrongNodeType } from './block/node/StrongNode'
export { FormulaNodeType } from './block/node/FormulaNode'
export { DecorationNodeType, DecorationType } from './block/node/DecorationNode'
export { CodeNodeType } from './block/node/CodeNode'
export { BlankNodeType } from './block/node/BlankNode'
export { UrlNodeType } from './block/node/UrlNode'
export { GoogleMapNodeType } from './block/node/GoogleMapNode'
export { InternalLinkNodeType } from './block/node/InternalLinkNode'
export { IconNodeType } from './block/node/IconNode'
export { HashTagNodeType } from './block/node/HashTagNode'
export { PlainNodeType } from './block/node/PlainNode'
