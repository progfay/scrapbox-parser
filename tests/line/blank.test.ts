/* eslint-disable no-irregular-whitespace */
import { describe, it, expect } from "vitest";
import { parse } from "../../src";

describe("blank", () => {
  it("Simple half-space blank", () => {
    expect(parse("[ ]", { hasTitle: false })).toMatchSnapshot();
  });

  it("Simple double-byte space blank", () => {
    expect(parse("[　]", { hasTitle: false })).toMatchSnapshot();
  });

  it("Simple tab blank", () => {
    expect(parse("[\t]", { hasTitle: false })).toMatchSnapshot();
  });

  it("Multi char blank", () => {
    expect(parse("[ 　 \t　\t ]", { hasTitle: false })).toMatchSnapshot();
  });

  it("Blank in the sentence", () => {
    expect(
      parse("sentence[ ]sentence", {
        hasTitle: false,
      }),
    ).toMatchSnapshot();
  });

  it("[] is not blank", () => {
    expect(parse("[]", { hasTitle: false })).toMatchSnapshot();
  });

  it("Blank in the [*** ]", () => {
    expect(parse("[*** [ ]]", { hasTitle: false })).toMatchSnapshot();
  });
});
