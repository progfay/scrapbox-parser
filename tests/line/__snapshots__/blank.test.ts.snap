// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`blank > [] is not blank 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "[]",
        "text": "[]",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`blank > Blank in the [*** ] 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "[*** ",
        "text": "[*** ",
        "type": "plain",
      },
      {
        "raw": "[ ]",
        "text": " ",
        "type": "blank",
      },
      {
        "raw": "]",
        "text": "]",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`blank > Blank in the sentence 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "sentence",
        "text": "sentence",
        "type": "plain",
      },
      {
        "raw": "[ ]",
        "text": " ",
        "type": "blank",
      },
      {
        "raw": "sentence",
        "text": "sentence",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`blank > Multi char blank 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "[ 　 	　	 ]",
        "text": " 　 	　	 ",
        "type": "blank",
      },
    ],
    "type": "line",
  },
]
`;

exports[`blank > Simple double-byte space blank 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "[　]",
        "text": "　",
        "type": "blank",
      },
    ],
    "type": "line",
  },
]
`;

exports[`blank > Simple half-space blank 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "[ ]",
        "text": " ",
        "type": "blank",
      },
    ],
    "type": "line",
  },
]
`;

exports[`blank > Simple tab blank 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "[	]",
        "text": "	",
        "type": "blank",
      },
    ],
    "type": "line",
  },
]
`;
